File: Interview Preparation Kit/Dictionaries and Hashmaps/CountTriplets.js
// https://www.hackerrank.com/challenges/count-triplets-1/problem
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');

    main();
});

function readLine() {
    return inputString[currentLine++];
}

// Complete the countTriplets function below.
function countTriplets(arr, r) {
    let first = new Map();
    let second = new Map();
    let triplets = 0;
    
    for (let i = 0; i < arr.length; i++) {
        let current = arr[i];
        if (current % r == 0) {
            let preceding = current / r;
            if (second.has(preceding)) {
                triplets += second.get(preceding);
            }
            if (first.has(preceding)) {
                if (second.has(current)) {
                    second.set(current, second.get(current) + first.get(preceding));
                }
                else {
                    second.set(current, first.get(preceding));
                }
            }
        }
        
        if (first.has(current)) {
            first.set(current, first.get(current) + 1);
        }
        else {
            first.set(current, 1);
        }
    }
    
    return triplets;
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const nr = readLine().replace(/\s+$/g, '').split(' ');

    const n = parseInt(nr[0], 10);

    const r = parseInt(nr[1], 10);

    const arr = readLine().replace(/\s+$/g, '').split(' ').map(arrTemp => parseInt(arrTemp, 10));

    const ans = countTriplets(arr, r);

    ws.write(ans + '\n');

    ws.end();
}

File: Interview Preparation Kit/Dictionaries and Hashmaps/CountTriplets.go
// https://www.hackerrank.com/challenges/count-triplets-1/problem
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

// Complete the countTriplets function below.
func countTriplets(arr []int64, r int64) int64 {
	first, second := make(map[int64]int64), make(map[int64]int64)
	triplets := int64(0)

	for _, current := range arr {
		if current%r == 0 {
			preceding := current / r
			val, exists := second[preceding]
			if exists {
				triplets += val
			}
			val, exists = first[preceding]
			if exists {
				_, exists = second[current]
				if exists {
					second[current] += val
				} else {
					second[current] = val
				}
			}
		}

		_, exists := first[current]
		if exists {
			first[current]++
		} else {
			first[current] = 1
		}
	}

	return triplets
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 16*1024*1024)

	stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
	checkError(err)

	defer stdout.Close()

	writer := bufio.NewWriterSize(stdout, 16*1024*1024)

	nr := strings.Split(strings.TrimSpace(readLine(reader)), " ")

	nTemp, err := strconv.ParseInt(nr[0], 10, 64)
	checkError(err)
	n := int32(nTemp)

	r, err := strconv.ParseInt(nr[1], 10, 64)
	checkError(err)

	arrTemp := strings.Split(strings.TrimSpace(readLine(reader)), " ")

	var arr []int64

	for i := 0; i < int(n); i++ {
		arrItem, err := strconv.ParseInt(arrTemp[i], 10, 64)
		checkError(err)
		arr = append(arr, arrItem)
	}

	ans := countTriplets(arr, r)

	fmt.Fprintf(writer, "%d\n", ans)

	writer.Flush()
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}
